packages/data/src/redux-store/metadata/actions.js

```diff
@@ -2,10 +2,10 @@
  * Returns an action object used in signalling that selector resolution has
  * started.
  *
- * @param {string} selectorName Name of selector for which resolver triggered.
- * @param {...*}   args         Arguments to associate for uniqueness.
+ * @param {string}    selectorName Name of selector for which resolver triggered.
+ * @param {unknown[]} args         Arguments to associate for uniqueness.
  *
- * @return {Object} Action object.
+ * @return {{ type: 'START_RESOLUTION', selectorName: string, args: unknown[] }} Action object.
```

Comment:
TypeScript features are really nice with action creators. Rather than effectively writing out the object twice, we use a const assertion and let type inference do the work. No need annotate the return type.

```ts
export function someAction( arg ) {
  return {
    type: 'SOME_ACTION',
    payload: arg,
  } as const; // <- "const assertion"
}
```

One thing to consider in the entire data layer is the use of readonly types. We could use mapped types like TypeScript [`Readonly<Type>`](https://www.typescriptlang.org/docs/handbook/utility-types.html#readonlytype) utility (shallowly sets `readonly`), [some `DeepReadonly` implementation](https://github.com/microsoft/TypeScript/issues/13923), or by adding `readonly` manually:

```suggestion
 * @return {{ readonly type: 'START_RESOLUTION', readonly selectorName: string, readonly args: readonly unknown[] }} Action object.
```

Considerations:

  - The types are more verbose
  - It may be applied inconsistently. There's room for error.
  - Doing it in the action creators themselves would require users to follow the same pattern.
  - This better reflects usage and expectations.
  - We may be able to apply readonly at the framework level in store creation/registration, benefitting more users of the library. This won't provide string literal `action.type` by default like const assertions can, but I don't think there's any way around that.

Reply:
So... which do we do? Convert the file to TS and use the const assertion or make the types readonly?

Reply:
This is a larger consideration than this PR, it's something we should keep in mind especially as we introduce types around `Store` objects.

_My preference_ is to use `as const` for action creators. It produces the best result with the lowest effort. A possible downside is that the documentation generator may not handle it well, I'm not sure and haven't tested.

Reply:
> My preference is to use as const for action creators. It produces the best result with the lowest effort.

Should we start with this approach and make changes if necessary?

Reply:
The documentation generator isn't able to infer return types (it doesn't actually run TypeScript compiler, just looks at the static code), they must be explicity declared, so using `as const` is hindered in that regard. Here these files are not affected by the documentation generator so we can use `as const` without any problem, but we might not be able to do so in all cases 🤷